### PR TITLE
Fix dehydrateState does not apply server side props

### PR DESCRIPTION
**Pull request by clgeoio**
Closes: https://github.com/blitz-js/blitz/issues/2601

### What are the changes and their implications?

The query that was setup in the dehydrated state did not have matching params (`null` vs `undefined`). Additionally, the SuperJSON deserialisation seemed to be causing some issues with the cached query being used on the SSR.

However, this change looks to undo: https://github.com/blitz-js/blitz/pull/2512

## Bug Checklist

- [x] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)

## Feature Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)
- [ ] Documentation added/updated (submit PR to [blitzjs.com repo](https://github.com/blitz-js/blitzjs.com))

### CODE REVIEW

File: packages/core/src/blitz-app-root.tsx [hunk 1]
```diff
@@ -110,7 +109,7 @@ export function withBlitzAppRoot(UserAppRoot: React.ComponentType<any>) {
     }, [])
 
     const dehydratedState = props.pageProps.dehydratedState
-      ? SuperJSON.deserialize(props.pageProps.dehydratedState)
+      ? props.pageProps.dehydratedState
```
Comment (Skn0tt) [inline packages/core/src/blitz-app-root.tsx:112]: It's interesting to me that removing the deserialisation helps in this case, since babel-plugin-superjson-next should be serializing `dehydratedState`: https://github.com/blitz-js/blitz/blob/a28216d067ff03bbd954a553275972e6f9340bc8/packages/babel-preset/src/index.ts#L16 🤔

Comment (Skn0tt) [reply packages/core/src/blitz-app-root.tsx]: I pushed an update in 98bdcf8, which should fix the behaviour while retaining SuperJSON support. The problem basically was that `SuperJSON.deserialize` was a totally wrong call - see the 98bdcf8's commit message for more context.

Comment (clgeoio) [reply packages/core/src/blitz-app-root.tsx]: Awesome, thanks for making that change!

I think I will add a test for some of the more complex types, like Map!

File: test/integration/queries/pages/dehydrated-state.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ import {Suspense} from "react"
 
 export const getServerSideProps: GetServerSideProps = async (ctx) => {
   const queryClient = new QueryClient()
-  const queryKey = getQueryKey(getDate, null)
+  const queryKey = getQueryKey(getDate, undefined)
```
[the remark this answers is not in the document]
Comment (clgeoio) [reply test/integration/queries/pages/dehydrated-state.tsx:14]: Interesting suggestion, I like it. It took me a while to figure it out and there was a small piece in the react-query docs that made me realize there was a mismatch.

Anyway we can make it easier for users to form the right code I am in support of :)

File: packages/core/src/blitz-app-root.tsx (hunk 1 above)
Comment (clgeoio) [reply packages/core/src/blitz-app-root.tsx]: Interestly @Skn0tt, if I swap out `getDate` with `getComplexTypes` and return something like:
```
export default async function getComplexTypes() {
  return {
    date: new Date(0),
    map: new Map().set("foo", "bar"),
    obj: {
      one: 1,
      child: {
        two: 2,
      },
    },
  }
}
```

When `map: new Map()` is in the returned object, the test fails:
```
 NoSuchElementError: no such element: Unable to locate element: {"method":"css selector","selector":"#content"}
      (Session info: headless chrome=92.0.4515.107)

      77 |     it("should work", async () => {
      78 |       const browser = await webdriver(context.appPort, "/dehydrated-state")
    > 79 |       const content = await browser.elementByCss("#content")
```

But when the `map` key is omitted from the return object, the test passes.

Any ideas here, should this be supported?

Comment (Skn0tt) [reply packages/core/src/blitz-app-root.tsx]: Yes, it should definitely be supported. That's some weird behaviour! Will look into it.

File: test/integration/queries/pages/dehydrated-state.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (flybayer) [reply test/integration/queries/pages/dehydrated-state.tsx:14]: Yeah I knew from the start we needed to make this better.

How about this?

```ts
import {prefetchQuery} from 'blitz'

export const getServerSideProps = async ctx => {

  await prefetchQuery(getData, undefined)

  return { props: {} }
}
```

And then we automatically add `new QueryClient()`, `dehydrate` etc via a babel plugin.

[the remark this answers is not in the document]
Comment (clgeoio) [reply test/integration/queries/pages/dehydrated-state.tsx:14]: A small suggestion would be to require that `dehydratedState` be passed into `props: {}` to avoid the edge case of a user declaring a prop `dehydratedState` and having overridden without them knowing. Perhaps we could rename the prop to `prefetchedQuery` for clarity and then transform in `blitz-root`?

```
export const getServerSideProps = async ctx => {
  const dehydratedState = await prefetchQuery(getData, undefined)
  return { props: {
    dehydratedState
  } }
}

[the remark this answers is not in the document]
Comment (flybayer) [reply test/integration/queries/pages/dehydrated-state.tsx:14]: The babel plugin can throw a helpful error or something in that case, right? I think having `dehydratedState` exposed is exposing unnecessary implementation details.

File: packages/core/src/blitz-app-root.tsx (hunk 1 above)
Comment (clgeoio) [reply packages/core/src/blitz-app-root.tsx]: Thank you @Skn0tt.
Let me know if I can lend a hand.
I'll park this PR until you've had a look 👍

Comment (mikealche) [reply packages/core/src/blitz-app-root.tsx]: Hey hey, I'm just starting to learn and poke around the blitz codebase, but I tried to debug this issue. I've found that after 98bdcf8, a possibility that the test is failing is that the query keys don't match.

The dehydrated state on `blitz-app-root`is: 
```
{ dehydratedState: 
   { mutations: [],
     queries: 
      [ { state: 
           { data: Wed Dec 31 1969 21:00:00 GMT-0300 (Argentina Standard Time),
             dataUpdateCount: 1,
             dataUpdatedAt: 1628558585243,
             error: null,
             errorUpdateCount: 0,
             errorUpdatedAt: 0,
             fetchFailureCount: 0,
             fetchMeta: null,
             isFetching: false,
             isInvalidated: false,
             isPaused: false,
             status: 'success' },
          queryKey: 
           [ '/api/queries/getDate',
             { json: null, meta: { values: [ 'undefined' ] } } ],
          queryHash: '["/api/queries/getDate",{"json":null,"meta":{"values":["undefined"]}}]' } ] } }
```

while if you log `getQueryKey(getDate, null)` in the client,you get: 
```
{ queryKey: [ '/api/queries/getDate', { json: null } ] }
```


If you look closely you'll see that the latter code is missing the `meta` part of the key: 
```
queryKey: 
           [ '/api/queries/getDate',
             { json: null, meta: { values: [ 'undefined' ] } } ]

vs

{ queryKey: [ '/api/queries/getDate', { json: null } ] }
```

I'm not sure what should be done to fix this, but I hope this helps and also I'm available if you need any helping hand @Skn0tt

**Comment by flybayer** [comment]
@Skn0tt what's the status on this PR?

**Comment by Skn0tt** [comment]
> @Skn0tt what's the status on this PR?

looking at it now, let's hope I find something :)
